### PR TITLE
Add lambda model and api updates

### PR DESCRIPTION
**Pull request by joe4dev**
## Contributions

* Add support for [runtime management controls](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-update.html) at API level
* Add support for [SnapStart](https://docs.aws.amazon.com/lambda/latest/dg/snapstart.html) at API level
* Add snapshot tests for SnapStart API
* Update all lambda-related snapshot tests due to model updates (only for `aws_validated` tests)
* Add platform flag to docker clients for `build_image`
* Fix Lambda tests on ARM hosts. This enables running Lambda tests on ARM hosts.
* Conditionally skip the concurrency tests if the AWS account has too low limits (improve error messages).

## Limitations

* Runtime management controls use a hard-coded sha256 value (`8eeff65f6809a3ce81507fe733fe09b835899b99481ba22fd75b5a7338290ec1`) instead of inspecting the Docker image.
* Rust builds on ARM (tested on M1 MacBook Pro) are slow (~5 minutes) due to emulation. See Makefile for potential performance improvements.
* The SnapStart tests are a little slow (~1-2min) against AWS

## Breaking changes

* This PR includes model changes in lambda and will break CloudPods for the new lambda provider

### CODE REVIEW

File: tests/integration/cloudformation/resources/test_lambda.py [hunk 1]
```diff
@@ -577,6 +587,14 @@ def wait_logs():
             "$..StartingPosition",
             "$..StateTransitionReason",
             "$..Topics",
+            # resource index mismatch due to SnapStart
```
Comment (joe4dev) [inline tests/integration/cloudformation/resources/test_lambda.py:590]: That is an example where a new feature changes the indexing of snapshots because it matches a resource transformer. It is not ideal that we need to disable these snapshot verifications for the old provider due to an unrelated change.

File: tests/integration/awslambda/test_lambda_api.py [hunk 2]
```diff
@@ -2412,7 +2434,15 @@ def test_lambda_provisioned_lifecycle(self, lambda_client, create_lambda_functio
...
     condition=is_old_provider,
-    paths=["$..RevisionId", "$..Policy.Statement", "$..PolicyName", "$..PolicyArn", "$..Layers"],
+    paths=[
+        "$..RevisionId",
+        "$..Policy.Statement",
+        "$..PolicyName",
+        "$..PolicyArn",
+        "$..Layers",
+        # mismatching resource index due to SnapStart
+        "$..Statement.Condition.ArnLike.'AWS:SourceArn'",
```
Comment (joe4dev) [inline tests/integration/awslambda/test_lambda_api.py:2444]: `:` break the JsonPath snapshot comparison and need to be enquoted using single quotes

File: tests/integration/awslambda/test_lambda.py [hunk 3]
```diff
@@ -171,12 +171,12 @@ def fixture_snapshot(snapshot):
...
             "$..Layers",
-            "$..SnapStart",  # FIXME
+            "$..SnapStart",
         ],
     )
 else:
     pytestmark = pytest.mark.skip_snapshot_verify(
-        paths=["$..CodeSize", "$..SnapStart"],  # FIXME
+        paths=["$..CodeSize"],  # FIXME
     )
```
Comment (dominikschubert) [inline tests/integration/awslambda/test_lambda.py:180]: we can actually just remove codesize as well and add it to the "global" transformers above

File: tests/integration/awslambda/test_lambda_api.py [hunk 4]
```diff
@@ -953,6 +959,7 @@ def test_publish_with_wrong_sha256(
         )
         snapshot.match("publish_result", publish_result)
 
+    @pytest.mark.aws_validated
```
Comment (dominikschubert) [inline tests/integration/awslambda/test_lambda_api.py:962]: good catch :+1:

File: tests/integration/awslambda/test_lambda_api.py [hunk 5]
```diff
@@ -1301,13 +1309,15 @@ def test_notfound_and_invalid_routingconfigs(
...
-        "$..SnapStart"
-    ]
-)
 class TestLambdaRevisions:
+    @pytest.mark.skip_snapshot_verify(
+        # The RuntimeVersionArn is currently a hardcoded id and therefore does not reflect the ARN resource update
+        # from python3.9 to python3.8 in update_function_configuration_response_rev5.
+        paths=[
+            "update_function_configuration_response_rev5..RuntimeVersionConfig.RuntimeVersionArn",
+            "get_function_response_rev6..RuntimeVersionConfig.RuntimeVersionArn",
```
Comment (dominikschubert) [inline tests/integration/awslambda/test_lambda_api.py:1318]:
```suggestion
            "$..RuntimeVersionConfig.RuntimeVersionArn",
```

or was there a particular reason for only including those?

File: tests/integration/awslambda/test_lambda_api.py (hunk 2 above)
Comment (dominikschubert) [reply tests/integration/awslambda/test_lambda_api.py:2444]: Maybe we should add some validation for this to avoid errors like this. Was this fairly obvious when debugging?

File: tests/integration/cloudformation/resources/test_lambda.py (hunk 1 above)
Comment (dominikschubert) [reply tests/integration/cloudformation/resources/test_lambda.py:590]: Agreed. I'm generally not super happy with the state of the resource transformer. I've added this as a reference in our internal issue tracker for snapshots

**Review by dominikschubert — APPROVED** [review]
LGTM! :partying_face: 

Great that you uncovered a lot of hidden issues that we'll need to tackle at some point. (e.g. tests missing the validated marker, rust compilation, resource transformer)

File: tests/integration/cloudformation/resources/test_lambda.py (hunk 1 above)
Comment (joe4dev) [reply tests/integration/cloudformation/resources/test_lambda.py:590]: Good to track that one. I think this might become a nasty source of flaky tests with automated test re-validation.

File: tests/integration/awslambda/test_lambda_api.py (hunk 2 above)
Comment (joe4dev) [reply tests/integration/awslambda/test_lambda_api.py:2444]: The error was pretty specific (see below) but the stack trace wasn't particularly helpful in identifying the issue because the hook is executed after the test.
It could be nice to give some context when raising (e.g., which snapshot match or which verify skip path).

```
>       raise JsonPathParserError('Parse error at %s:%s near token %s (%s)'
                                  % (t.lineno, t.col, t.value, t.type))
E       jsonpath_ng.exceptions.JsonPathParserError: Parse error at 1:34 near token : (:)

.venv/lib/python3.10/site-packages/jsonpath_ng/parser.py:81: JsonPathParserError
```

File: tests/integration/awslambda/functions/common/endpointinjection/provided/Makefile [hunk 6]
```diff
@@ -1,8 +1,33 @@
+DOCKER_PLATFORM ?= linux/amd64
+# Using a sha256 hash breaks on multi-platform hosts (e.g., ARM Macs): https://github.com/moby/moby/issues/43188
+DOCKER_RUST_IMAGE ?= rust:1.67.1
+
+# x86 cross-build on ARM hosts are killed due to memory issues: https://github.com/rust-lang/cargo/issues/10781#issuecomment-1351670409
+# Slow workaround takes ~5 to build on an M1 MacBook Pro: --config net.git-fetch-with-cli=true
```
[the remark this answers is not in the document]
Comment (joe4dev) [reply tests/integration/awslambda/functions/common/endpointinjection/provided/Makefile]: yup, fixed

File: tests/integration/awslambda/test_lambda.py (hunk 3 above)
Comment (joe4dev) [reply tests/integration/awslambda/test_lambda.py]: Good catch 👍 
I just removed it because `CodeSize` is already part of the `lambda_api()` transformer.

File: tests/integration/awslambda/test_lambda_api.py (hunk 4 above)
Comment (joe4dev) [reply tests/integration/awslambda/test_lambda_api.py:962]: Thanks.
I filtered by `-m aws_validated`, hence all lambda tests with any drift should be marked now 🟢

File: tests/integration/awslambda/test_lambda_api.py (hunk 5 above)
Comment (joe4dev) [reply tests/integration/awslambda/test_lambda_api.py:1318]: I still wanted the validation that the `RuntimeVersionArn` does not change between revisions and not ignore it completely for the entire test
